### DISCUSSION TITLE
Add a way to intercept outgoing http requests

### DISCUSSION THREAD

**Discussion by smaznet** [telegraf/telegraf#1267] [Ideas] [answered]
I have a suggestion
Please add a way to middlewar outgoing requests

Something like this
``` js
bot.telegram.use(async (method,data,next)=>{
    let result = await next(method,data);
});
```

This is useful in some cases:

1. if you want save outgoing messages to database
2. modify some requests
3. if you want run some requests with delay
4. When you want send some requests which their flooding fast like sendContact
You can limit requests

**Comment by wojpawlik** [comment]
Currently, this can be safely achieved by replacing `ctx.telegram.callApi` in every request (see https://github.com/KnightNiwrem/telegraf-throttler/blob/master/src/index.ts), but I agree that it'd be nice to have a proper API for it.

**Comment by smaznet** [comment]
Yes i know and I'm using this method

**Comment by wojpawlik** [comment]
This will be a <ins>significant</ins> internal change (I'd expect it to ship in ~~4.2~~ <ins>5.0</ins>).

We will need a proposal, so here's mine. To keep the discussion manageable, please post any concerns about my proposal as a reply to this, and any alternatives as a standalone comment, so it can be discussed separately.

<hr>

`ApiClient` already contains a lot of complexity. Let's wrap it instead of adding more.

```ts
interface ApiClientInterface {
  callApi: <M extends keyof Telegram>(
    method: M,
    payload: Opts<M>,
    { signal, response }: ApiClient.CallApiOptions = {}
  ) => Promise<ReturnType<Telegram[M]>>
}

type Interceptor = (client: ApiClientInterface) => ApiClientInterface
```

`ApiClient` would ignore `response`, but an interceptor could use it to perform `webhookReply`.

It'd be great if some way to access response headers (`Date` to be specific ) was introduced.

**Breaking**: `Telegram` would need to accept and contain `ApiClientInterface` rather than accept token and inherit from `ApiClient`.

**Breaking**: that breaks `Telegram::getFileLink`, so it needs to be replaced with something. cc @AndrewLaneX

This decoupling would allow to write `MockApiClient` (#224).

`bot.intercept(interceptor)` would apply the interceptor to `bot.telegram` and all future `ctx.telegram`s.

**Comment by smaznet** [comment]
Hi
I thought about this and i got a simple idea

we can do it like this
(Function names is only a suggestion)
``` js
class ApiClient{

   setInterceptor(interceptorFunc){
    this.interceptor=interceptorFunc;
   }

// current call api method should be renamed to this
   _callApi(method,data){
        ...
   }

  callApi(method,data){
   if(this.interceptor){
      return this.interceptor(method,data,this._callApi)
    }else{
      return this._callApi(method,data);
   }
  }
}
```
and can be used like this
``` js
bot.telegram.setInterceptor(async(method, data,callApi)=>{
    // edit request if needed
    let result = await callApi(method,data);
    // Do anything with result
    return result;
})
```

**Reply by wojpawlik** [reply]
This only allows one interceptor, and adds complexity to `ApiClient`.

**Reply by smaznet** [reply]
Yes
why someone wants more interceptors?

**Reply by smaznet** [reply]
Oh That's bad idea 
Telegraf creates ApiClient on each incoming update and this code won't work

**Reply by wojpawlik** [reply]
> why someone wants more interceptors?

I could imagine wanting to use `telegraf-throttler` and auto `webhookReply` alongside my bot-specific, custom interceptor.

**Comment by KnorpelSenf** [comment]
In case anyone is still interested in this, an alternative package that I created (called grammY) features exactly what this discussion is asking for.

https://grammy.dev

**Comment by wojpawlik** [comment, marked as answer]
#1704
